tests/unit/test_index_contents_hierarchy.py

```diff
@@ -0,0 +1,712 @@
+                    whitespace_count=0, reference_value=(value_1 := "dir_1")
+                ),
+                item_2 := factories.IndexParsedListItemFactory(
+                    whitespace_count=1, reference_value=(value_2 := f"{value_1}/dir_2")
+                ),
+                item_3 := factories.IndexParsedListItemFactory(
+                    whitespace_count=2, reference_value=(value_3 := f"{value_2}/file_3.md")
+                ),
+            ),
+            (create_dir, create_dir, create_file),
+            (
+                factories.IndexContentsListItemFactory(
+                    hierarchy=1,
+                    reference_title=item_1.reference_title,
+                    reference_value=value_1,
+                    rank=item_1.rank,
+                ),
+                factories.IndexContentsListItemFactory(
+                    hierarchy=2,
+                    reference_title=item_2.reference_title,
+                    reference_value=value_2,
+                    rank=item_2.rank,
+                ),
+                factories.IndexContentsListItemFactory(
+                    hierarchy=3,
+                    reference_title=item_3.reference_title,
+                    reference_value=value_3,
+                    rank=item_3.rank,
+                ),
+            ),
+            id="single directory single nested directory single file in nested directory",
+        ),
+        pytest.param(
+            (
+                item_1 := factories.IndexParsedListItemFactory(
+                    whitespace_count=0, reference_value=(value_1 := "dir_1")
+                ),
+                item_2 := factories.IndexParsedListItemFactory(
+                    whitespace_count=1, reference_value=(value_2 := f"{value_1}/dir_2")
+                ),
+                item_3 := factories.IndexParsedListItemFactory(
+                    whitespace_count=2, reference_value=(value_3 := f"{value_2}/dir_3")
+                ),
+            ),
+            (create_dir, create_dir, create_dir),
+            (
+                factories.IndexContentsListItemFactory(
+                    hierarchy=1,
+                    reference_title=item_1.reference_title,
+                    reference_value=value_1,
+                    rank=item_1.rank,
+                ),
+                factories.IndexContentsListItemFactory(
+                    hierarchy=2,
+                    reference_title=item_2.reference_title,
+                    reference_value=value_2,
+                    rank=item_2.rank,
+                ),
+                factories.IndexContentsListItemFactory(
+                    hierarchy=3,
+                    reference_title=item_3.reference_title,
+                    reference_value=value_3,
+                    rank=item_3.rank,
+                ),
+            ),
+            id="single directory single nested directory single directory in nested directory",
```

Comment:
Perhaps we should also test the case `dir_1`, `dir_1/file_2.md`, `dir_3` and `dir_3/file_4.md` ?

Reply:
yeh I think that is a fair test, should be added on next push